src/Common/src/Microsoft/Win32/SafeHandles/NtlmSecuritySafeHandles.cs

```diff
@@ -0,0 +1,271 @@
+    /// </summary>
+    internal sealed class SafeNtlmKeyHandle : SafeHandle
+    {
+        private uint _sequenceNumber;
+        private readonly bool _isSealingKey;
+        private readonly byte[] _digest;
+        private SafeEvpCipherCtxHandle _cipherContext;
+
+        // From MS_NLMP SIGNKEY at https://msdn.microsoft.com/en-us/library/cc236711.aspx
+        private const string s_keyMagic = "session key to {0}-to-{1} {2} key magic constant\0";
+        private const string s_client = "client";
+        private const string s_server = "server";
+        private const string s_signing = "signing";
+        private const string s_sealing = "sealing";
+
+        public SafeNtlmKeyHandle(byte[] key, bool isClient, bool isSealingKey)
+            : base(IntPtr.Zero, true)
+        {
+            string keyMagic = string.Format(s_keyMagic, isClient ? s_client : s_server,
+                    isClient ? s_server : s_client, isSealingKey ? s_sealing : s_signing);
+
+            using (IncrementalHash incremental = IncrementalHash.CreateHash(HashAlgorithmName.MD5))
+            {
+                 incremental.AppendData(key);
+                 incremental.AppendData(Encoding.UTF8.GetBytes(keyMagic));
+                 _digest = incremental.GetHashAndReset();
+            }
+
+            _isSealingKey = isSealingKey;
+            if (_isSealingKey)
+            {
+                _cipherContext = Interop.Crypto.EvpCipherCreate(Interop.Crypto.EvpRc4(), _digest, null, 1);
+            }
+        }
+
+        public override bool IsInvalid
+        {
+            get { return true; }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && (null != _cipherContext))
+            {
+                _cipherContext.Dispose();
+                _cipherContext = null;
+            }
+            base.Dispose(disposing);
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            return true;
+        }
+
+        public byte[] Sign(SafeNtlmKeyHandle sealingKey, byte[] buffer, int offset, int count)
+        {
+            Debug.Assert(!_isSealingKey, "Cannot sign with sealing key");
+            Debug.Assert(offset >= 0 && offset < buffer.Length, "Cannot sign with invalid offset " + offset);
+            Debug.Assert((count + offset) <= buffer.Length, "Cannot sign with invalid count " + count);
+
+            // reference for signing a message: https://msdn.microsoft.com/en-us/library/cc236702.aspx
+            const uint Version = 0x00000001;
+            const int ChecksumOffset = 4;
+            const int SequenceNumberOffset = 12;
+            const int HMacDigestLength = 8;
+
+            byte[] output = new byte[Interop.NetSecurity.MD5DigestLength];
+            Array.Clear(output, 0, output.Length);
+            MarshalUint(output, 0, Version); // version
+            MarshalUint(output, SequenceNumberOffset, _sequenceNumber);
```

Comment:
Rather than writing your own marshaling method, how about:
```C#
using (byte* outputPtr = &output)
{
    Marshal.WriteInt32((IntPtr)outputPtr, 0,  (int)Version);
    Marshal.WriteInt32((IntPtr)outputPtr, SequenceNumberOffset, (int)_sequenceNumber);
}
```
?
I can understand the desire to avoid unsafe code like this.  But I wanted to highlight that such functions are available if you want to use them (it's too bad BitConverter doesn't have overloads that write into an existing array rather than allocating new ones, cc: @KrzysztofCwalina).